web-admin/src/features/scheduled-reports/query-mapper.ts

```diff
@@ -0,0 +1,277 @@
+ * Reports manually written through file artifacts won't have the UI to feed the url state.
+ * Hence we are building the state from the query args in the report.
+ */
+export function parseReportQuery(
+  reportResource: V1Resource,
+  executionTime: string,
+): Readable<ParsedReportQuery> {
+  if (!reportResource?.report?.spec?.queryName)
+    return readable({
+      ready: false,
+    });
+
+  let metricsViewName: string = "";
+  const req: ReportQueryRequest = correctRequest(
+    JSON.parse(reportResource.report.spec.queryArgsJson),
+  );
+  let queryMapper: (args: QueryMapperArgs<ReportQueryRequest>) => void;
+
+  // get metrics view name and the query mapper function based on the query name.
+  switch (reportResource.report.spec.queryName) {
+    case "MetricsViewAggregation":
+      metricsViewName = (req as V1MetricsViewAggregationRequest).metricsView;
+      queryMapper = mapMetricsViewAggregationRequest;
+      break;
+    case "MetricsViewToplist":
+      metricsViewName = (req as V1MetricsViewToplistRequest).metricsViewName;
+      queryMapper = mapMetricsViewToplistRequest;
+      break;
+    case "MetricsViewRows":
+      metricsViewName = (req as V1MetricsViewRowsRequest).metricsViewName;
+      queryMapper = mapMetricsViewRowsRequest;
+      break;
+    case "MetricsViewTimeSeries":
+      metricsViewName = (req as V1MetricsViewTimeSeriesRequest).metricsViewName;
+      queryMapper = mapMetricsViewTimeSeriesRequest;
+      break;
+    case "MetricsViewComparison":
+      metricsViewName = (req as V1MetricsViewComparisonRequest).metricsViewName;
+      queryMapper = mapMetricsViewComparisonRequest;
+      break;
+  }
+
+  if (!metricsViewName) {
+    // error state
+    return readable({
+      ready: true,
+      error:
+        "Failed to find metrics view name. Please check the format of the report.",
+    });
+  }
+
+  return derived(
+    [
+      useMetaQuery(get(runtime).instanceId, metricsViewName),
+      // TODO: handle non-timestamp dashboards
+      createQueryServiceMetricsViewTimeRange(
+        get(runtime).instanceId,
+        metricsViewName,
+        {},
+      ),
+    ],
+    ([metricsViewResource, timeRangeSummary]) => {
+      if (!metricsViewResource.data || !timeRangeSummary.data)
+        return {
+          ready: false,
+        };
+
+      if (metricsViewResource.error || timeRangeSummary.error) {
+        // error state
+        return {
+          ready: true,
+          error:
+            metricsViewResource.error?.message ??
+            timeRangeSummary.error?.message,
+        };
+      }
+
+      initLocalUserPreferenceStore(metricsViewName);
+      const dashboard = getDefaultMetricsExplorerEntity(
+        metricsViewName,
+        metricsViewResource.data,
+        timeRangeSummary.data,
+      );
+      queryMapper({
+        dashboard,
+        req,
+        metricsView: metricsViewResource.data,
+        timeRangeSummary: timeRangeSummary.data.timeRangeSummary,
+        executionTime,
```

Comment:
Is it purposeful that the function returns `void`? Might be more clear if it returned the dashboard state object

Reply:
It updated the dashboard in place. It felt like an overkill to make it copy again and return it, since we are creating the dashboard object before passing it to the mapper.